docs/development/when_to_rebase.rst

```diff
@@ -0,0 +1,45 @@
+*********************************
+When to rebase and squash commits
+*********************************
+
+This page describes recommendations for when to rebase pull requests and when to
+combine/squash commits.
+
+When to remove or combine/squash commits
+========================================
+
+Pull requests **must** be rebased and at least partially squashed (but not
+necessarily squashed to a single commit) if large (approximately >10KB)
+non-source code files (e.g. images, data files, etc.) are added and then removed
+or modified in the PR commit history (The squashing should remove all but the
+last addition of the file to not use extra space in the repository).
+
+Combining/squashing commits is **encouraged** but not required where the number
+of commits is excessive for the changes made.  The definition of 'excessive' is
+subjective, but would never apply for 5 or fewer commits even for a trivial
+change.  Reviewers should use their discretion in suggesting to combine/squash
+commits.  Be mindful of maintaining a welcoming environment and be helpful with
+advice, especially for new contributors.
+
+As an example, for a change affecting < 10 lines of source code and including a
+changelog entry, more than 5 commits would be excessive. For a larger pull
+request adding significant functionality, 20 or 30 commits may on the other hand
+be appropriate/acceptable.
+
+When to rebase
+==============
+
+Pull requests **must** be rebased (but not necessarily squashed to a single
+commit) if at least one of the following conditions is met:
+
+* There are conflicts with master
+* There are merge commits from upstream/master in the PR commit history (merge
+  commits from PRs to the user's fork are fine)
+* There are commit messages include offensive language or violate the code of
```

Comment:
How does rebasing (with no squashing) get rid of the offensive language or the merge commits? The "not necessarily squashed" does not apply to those.

Reply:
I agree, that should be put in the "squashed" section.

Reply:
rebase does let you edit commit messages as well...

Reply:
I'm sure all of the CoC violating cases will be asked for to deal with, rather than the original committer reads this docs and changes their behaviour.